### PR TITLE
WIP: Add DRYrun support

### PR DESCRIPTION
**Pull request by johncmckim**
Pull Request Template ------------------------

##### Status:
In Development

##### Description:
Add's a dryRun flag as described in #1496

##### Todos:
- [x] Prevent functions being uploaded to S3
- [x] Prevent cloudformation stacks being updated
- [x] Write cloudformation template to disk
- [x] Write tests
- [x] Consider whether a separate command is more appropriate - it isn't

### CODE REVIEW

**Comment by pmuens** [comment]
Thanks for the hard work you've put into here @johncmckim!

IMHO as discussed in #1496 we should consider to switch to an own plugin as it encapsulates the functionality and implements dryrun in a more pluginish fashion (and not spread across multiple places).

**Comment by flomotlik** [comment]
@pmuens @johncmckim I like the --dry-run option as it is pretty much used in this way in many tools (e.g. the AWS CLI)

The only problem I see at the moment is that in the package plugin we also have the cleanup task which removes the whole .serverless folder. So you can't actually get access to the artefacts after `serverless deploy --dryRun`

Imho there are 2 options.

1. We completely remove cleanup (which means the .serverless folder will be growing quite a bit)
2. We ignore cleanup when the --dryRun option is set (probably better for now)

Thoughts?

**Comment by johncmckim** [comment]
@flomotlik @pmuens I'm for the new plugin rather than a command option. With this implementation the flag doesn't actually dry run a deploy, it just builds artefacts. I think that makes the flag a little misleading. That said, I'll go with whatever you decide.

In regards to the cleanup, when I ran this locally the `.serverless` folder wasn't getting cleaned up. But I may have missed something. I'll check.

What if we ran a cleanup before deploy rather than after. That way the `.serverless` folder doesn't bloat and the user can see the artefacts after a deploy. As part of this, I'd write the cf template to disk every time not just on `dryRun`. In the past I've wanted to see the generated cf template on a failed deploy but it wasn't accessible.

**Comment by flomotlik** [comment]
>@flomotlik @pmuens I'm for the new plugin rather than a command option. With this implementation the flag doesn't actually dry run a deploy, it just builds artefacts. I think that makes the flag a little misleading. That said, I'll go with whatever you decide.

How would this plugin/command work? It would simply package the artefacts, or what specifically would happen as part of the dryRun (that is different from what is happening here). Maybe we should rename `--dryRun` to something that says "We're going to build artefacts, but we're not going to deploy them"

>What if we ran a cleanup before deploy rather than after. That way the .serverless folder doesn't bloat and the user can see the artefacts after a deploy. As part of this, I'd write the cf template to disk every time not just on dryRun. In the past I've wanted to see the generated cf template on a failed deploy but it wasn't accessible.

That sounds good to me.

**Comment by pmuens** [comment]
> What if we ran a cleanup before deploy rather than after. That way the .serverless folder doesn't bloat and the user can see the artefacts after a deploy. As part of this, I'd write the cf template to disk every time not just on dryRun. In the past I've wanted to see the generated cf template on a failed deploy but it wasn't accessible.

I like that one! Really makes sense (some kind of the simplest `dry run` implementation possible).

However I would still vote for an extra plugin which encapsulates everything. Otherwise we'll end up with so many code spread across multiple files (`if` checks if the `dryrun` option is set to true) which makes it hard to maintain, and extend later on.
The plugin is also a good way to communicate that feature as it's listed as a core plugin with an own command (something like `serverless plan`, or `serverless deploy simulate`) instead of a hidden option which might be hard to find in the docs.

**Comment by flomotlik** [comment]
Ok so basically what we're talking about is actually not necessarily dryRun (here at least) but more `serverless package` to create the artefacts. @johncmckim is that also your idea of it or how would dryRun be different for you?

**Comment by flomotlik** [comment]
One of the reasons why I like this not being a separate plugin is that its future proof. Basically if we update the way deployments happen we need to make dryRun work as well. If its a separate plugin or functionality we might not and it doesn't work correctly any more.

**Comment by johncmckim** [comment]
@flomotlik you were right, it was still cleaning up resources. I didn't see it when I was testing as I hadn't setup my `AWS_PROFILE`.

@pmuens @flomotlik just changed clean to occur before deploying. I've added a new lifecycle event `deploy:clean` that occurs before `deploy:initializeResources`. However, I can change it to use `before:deploy:initializeResources` if you would prefer that.

In relation to the command vs argument, `serverless package` is what I was thinking. However, the maintainability issue is a fair point. 

I think I just don't like the `--dryRun` terminology. This is probably because of my experience with `boto` and the aws cli. In both those systems dry run does this for `ec2`

> DryRun (boolean) -- Checks whether you have the required permissions for the action, without actually making the request, and provides an error response. If you have the required permissions, the error response is DryRunOperation . Otherwise, it is UnauthorizedOperation .

Basically I think `--dryRun` implies that we're doing more just packaging the artefacts and skipping the deploy.

I do like `--plan` as to me at least it implies that we're going to get the artefacts ready but do nothing with them. Terraform behaves in this way.

Thoughts? I'm open to either a new plugin, `--dryRun` or `--plan`. In the end, they'll all do the job. It's just deciding on the best way to express it.

File: lib/plugins/aws/deploy/lib/createStack.js [hunk 1]
```diff
@@ -2,8 +2,18 @@
 
 const BbPromise = require('bluebird');
 const async = require('async');
+const path = require('path');
 
 module.exports = {
+  writeCreateTemplateToDisk() {
+    const cfTemplateFilePath = path.join(this.serverless.config.servicePath,
+      '.serverless', `cf-template-create-${(new Date).getTime().toString()}.json`);
```
Comment (johncmckim) [inline lib/plugins/aws/deploy/lib/createStack.js]: @pmuens @flomotlik I added a date to the name to keep it inline with the zip package.

Do you think a date is necessary here? It makes it hard to test and if the `.serverless` folder is being cleaned up it's probably unnecessary.

**Comment by flomotlik** [comment]
Imho `--plan` is still a bit hard to understand, because it doesn't really say what planning actually is in this context. What about `--no-deploy`? It clearly states what is going to happen. Looks a bit silly with `serverless deploy --no-deploy` but at least its very understandable. Another option would be `--package-only`, although the whole process might be doing more than just packaging things up so this could be a bit misleading as well.

**Comment by flomotlik** [comment]
lets go with `--no-deploy` :D. We can easily change it later if we have a better idea

**Comment by johncmckim** [comment]
I've updated it to `--noDeploy` which isn't as nice as `--no-deploy`. I was having a few issues with `minimist`.

To use `--no-deploy` I need to create a `deploy` option and `minimist` will convert `--deploy` and `--no-deploy` to `true` and `false`. However, since there is no ability to pass defaults to `minimist` the default is false.

I could solve that with `_.isNil` checks everywhere. However, I wasn't sure that it is the best solution. I could add a `default` property to plugin command options. This could then be passed into `minimist`.

What would you prefer?

**Comment by pmuens** [comment]
@johncmckim thanks for the updates! 💃 

I would do the simple solution / the less "hacky" one and vote for `--noDeploy` for now.

Will look into the PR in detail soon.
At a first glance it looks really good 👍 
The only thing I'm thinking about is the introduction of the new `clean` lifecycle which could be replaced with `before:deploy:deploy`. The "problem" with adding a new lifecycle is that we need to support those as plugin authors depend on it and we want to understand the other cloud providers in detail so that we have a very general and minimalistic set of lifecycle events every cloud provider would use.

Writing the CloudFormation template to disc might be related to this issue:
https://github.com/serverless/serverless/issues/1740

So this PR could be a good preparation for #1740 🎉

**Comment by flomotlik** [comment]
@pmuens I actually like the clean step. Imho its something that sometimes needs to happen at the beginning of a deployment, so Imho introducing a separate step makes it easier for other providers in the future to integrate there. Also in favor of --noDeploy. I also tried it locally and it properly created the files and didn't deploy.

From my perspective this is GTM, @pmuens could you review the code as well and merge in case the code is fine for you.

File: lib/plugins/aws/deploy/lib/createStack.js (hunk 1 above)
Comment (flomotlik) [inline lib/plugins/aws/deploy/lib/createStack.js]: Imho timestamp is good here so it doesn't overwrite anything in the future in case we change this implementation a bit and keep more things around.

**Comment by pmuens** [comment]
Alright! Will look into this and merge once fully reviewed.

Thanks @johncmckim for your work on this one!

File: lib/plugins/aws/deploy/tests/createStack.js [hunk 2]
```diff
@@ -151,6 +151,35 @@ describe('createStack', () => {
       });
     });
 
+    it('should resolve if dry run', () => {
```
Comment (pmuens) [inline lib/plugins/aws/deploy/tests/createStack.js]: Could you update this test description (`should resolve if no deploy`)?

File: lib/plugins/aws/deploy/tests/updateStack.js [hunk 3]
```diff
@@ -152,14 +152,37 @@ describe('updateStack', () => {
   });
 
   describe('#updateStack()', () => {
+    it('should resolve if dry run', () => {
```
Comment (pmuens) [inline lib/plugins/aws/deploy/tests/updateStack.js]: Could you update this test description?

File: lib/plugins/aws/deploy/tests/uploadDeploymentPackage.js [hunk 4]
```diff
@@ -200,6 +200,31 @@ describe('uploadDeploymentPackage', () => {
   });
 
   describe('#uploadDeploymentPackage()', () => {
+    it('should resolve if dry run', () => {
```
Comment (pmuens) [inline lib/plugins/aws/deploy/tests/uploadDeploymentPackage.js]: Could you update this test description?

File: lib/plugins/deploy/deploy.js [hunk 5]
```diff
@@ -8,6 +8,7 @@ class Deploy {
       deploy: {
         usage: 'Deploy Service.',
         lifecycleEvents: [
+          'clean',
```
Comment (pmuens) [inline lib/plugins/deploy/deploy.js]: I would propose to rename this to `cleanup` instead of `clean`

**Comment by pmuens** [comment]
@johncmckim just reviewed it and added a few comments (mainly some renamings).

Furthermore some documentation about this new feature would be great (some description on how to use it here: https://github.com/serverless/serverless/tree/master/lib/plugins/deploy and the new lifecycle here: https://github.com/serverless/serverless/tree/master/lib/plugins/deploy#provided-lifecycle-events)

After those changes it's GTM from my side.

Great work :tada: ❤️

File: lib/plugins/package/index.js [hunk 6]
```diff
@@ -22,7 +22,7 @@ class Package {
         .then(this.validate)
         .then(this.zipService),
 
-      'after:deploy:deploy': () => BbPromise.bind(this)
+      'deploy:clean': () => BbPromise.bind(this)
```
Comment (pmuens) [inline lib/plugins/package/index.js]: Could you move this above `deploy:createDeploymentArtifacts` so that correct order the hooks run through can be seen without looking into the `deploy` plugin.

**Comment by pmuens** [comment]
@johncmckim great! Thanks for the updates 👍 . Will pull, test and merge it today!

File: lib/plugins/aws/deploy/index.js [hunk 7]
```diff
@@ -35,7 +35,13 @@ class AwsDeploy {
         .then(this.uploadDeploymentPackage)
         .then(this.deployFunctions)
         .then(this.updateStack)
-        .then(() => this.serverless.cli.log('Deployment successful!')),
+        .then(() => {
+          const msg = this.options.noDeploy ?
+            'Deployment successful!' :
+            'Did not deploy due to --noDeploy';
```
Comment (eahefnawy) [inline lib/plugins/aws/deploy/index.js]: We should probably remove the "Deployment successful" message since deployment never really happened.

Comment (johncmckim) [inline lib/plugins/aws/deploy/index.js]: Ah I think I've got my messages the wrong way around. That would be confusing. I can update later today.

**Comment by nicka** [comment]
Another reason I really like this dry-run/no-deploy is that it allows for a CF diff plugin. First generate the new template without deploying and diffing against the running CF stack. This allows you to be more confident about what CF is going to perform during the deployment.

**Comment by johncmckim** [comment]
@nicka although it's a separate issue I've been wondering about that too. Whether serverless could or should use CloudFormation Change Sets and the separation of build and deploy.

With a few small changes you could support deploying pre-built artefacts. This would be helpful for projects like [LambCI](https://github.com/lambci/lambci) where people deploy the project into their own environment.

**Comment by nicka** [comment]
@johncmckim Not sure about the CloudFormation Change Sets(maybe haven't used it enough). This could even be it's own plugin, to build and to deploy. For me personally simply seeing the diff between two templates would already be a useful addition. Although it might take some more skill to predict what will happen based on the diff, it could also help new users to get used to CF.

**Comment by flomotlik** [comment]
thanks @johncmckim for this!

**Comment by thenikso** [comment]
Looking forward for beta2

question: will I be able with `--noDeploy` to see the merged/final CloudFormation configuration? It would be extremely useful to understand if I'll be deploying exactly what I'd like to

**Comment by ozbillwang** [comment]
Seems this PR is gone, I can't find the option of `noDeploy` now. 

Will `sls` still support this feature?
